frontend_services/opengl_glfw/OpenGL_GLFW_Service.cpp

```diff
@@ -0,0 +1,651 @@
+    // of another Window/OpenGL context. So this handle is private, but can be used by other GLFW service instances too.
+};
+
+void initSharedContext(SharedData& context) {
+    context.borrowed_glfwContextWindowPtr = nullptr; // stays null since nobody shared his GL context with us
+}
+
+// indirection to not spam header file with GLFW inclucde
+#define that static_cast<OpenGL_GLFW_Service*>(::glfwGetWindowUserPointer(wnd))
+
+// keyboard events
+void outer_glfw_onKey_func(GLFWwindow* wnd, int key, int scancode, int action, int mods) {
+    that->glfw_onKey_func(key, scancode, action, mods);
+}
+void outer_glfw_onChar_func(GLFWwindow* wnd, unsigned int codepoint) { that->glfw_onChar_func(codepoint); }
+
+// mouse events
+void outer_glfw_onMouseButton_func(GLFWwindow* wnd, int button, int action, int mods) {
+    that->glfw_onMouseButton_func(button, action, mods);
+}
+void outer_glfw_onMouseCursorPosition_func(GLFWwindow* wnd, double xpos, double ypos) {
+	// cursor (x,y) position in screen coordinates relative to upper-left corner
+    that->glfw_onMouseCursorPosition_func(xpos, ypos);
+}
+void outer_glfw_onMouseCursorEnter_func(GLFWwindow* wnd, int entered) {
+    that->glfw_onMouseCursorEnter_func(entered == GLFW_TRUE);
+}
+void outer_glfw_onMouseScroll_func(GLFWwindow* wnd, double xoffset, double yoffset) {
+    that->glfw_onMouseScroll_func(xoffset, yoffset);
+}
+
+const char* outer_glfw_getClipboardString(void* user_data) {
+	return glfwGetClipboardString(reinterpret_cast<GLFWwindow*>(user_data));
+}
+void outer_glfw_setClipboardString(void* user_data, const char* string) {
+	glfwSetClipboardString(reinterpret_cast<GLFWwindow*>(user_data), string);
+}
+
+// window events
+void outer_glfw_onWindowSize_func(GLFWwindow* wnd, int width /* in screen coordinates of the window */, int height) {
+    that->glfw_onWindowSize_func(width, height);
+}
+void outer_glfw_onWindowFocus_func(GLFWwindow* wnd, int focused) { that->glfw_onWindowFocus_func(focused == GLFW_TRUE); }
+void outer_glfw_onWindowShouldClose_func(GLFWwindow* wnd) { that->glfw_onWindowShouldClose_func(true); }
+void outer_glfw_onWindowIconified_func(GLFWwindow* wnd, int iconified) { that->glfw_onWindowIconified_func(iconified == GLFW_TRUE); }
+void outer_glfw_onWindowContentScale_func(GLFWwindow* wnd, float xscale, float yscale) {
+    that->glfw_onWindowContentScale_func(xscale, yscale);
+}
+// void outer_glfw_WindowPosition_func(GLFWwindow* wnd, int xpos, int ypos) { that->glfw_WindowPosition_func(xpos, ypos); }
+void outer_glfw_onPathDrop_func(GLFWwindow* wnd, int path_count, const char* paths[]) {
+    that->glfw_onPathDrop_func(path_count, paths);
+}
+
+// framebuffer events
+void outer_glfw_onFramebufferSize_func(GLFWwindow* wnd, int widthpx, int heightpx) {
+    that->glfw_onFramebufferSize_func(widthpx, heightpx);
+}
+
+} // namespace
+
+// helpers to simplify data access
+#define m_data (*m_pimpl)
```

Comment:
Maybe this extense use of macros is not so nice.
The code maybe is shorter, but much harder to read. It may be unclear where this variable comes from, its counter intuitive, makes it harder for IDE code scanning and especially the defines are not bound to the current scope.